compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java

```diff
@@ -544,64 +586,102 @@ private void checkUniqueness(BLangConstant constant) {
+    private BFiniteType createFiniteType(BConstantSymbol constantSymbol, BLangExpression expr) {
+        BTypeSymbol finiteTypeSymbol = Symbols.createTypeSymbol(SymTag.FINITE_TYPE, constantSymbol.flags, Names.EMPTY,
+                                                                constantSymbol.pkgID, null, constantSymbol.owner,
+                                                                constantSymbol.pos, VIRTUAL);
         BFiniteType finiteType = new BFiniteType(finiteTypeSymbol);
         finiteType.addValue(expr);
         return finiteType;
     }
 
-    private BType checkType(BLangExpression expr, BLangConstant constant, Object value, BType type, Location pos,
-                            Object constValue) {
+    private BType checkType(BLangExpression expr, BConstantSymbol constantSymbol, Object value, BType type,
+                            Location pos, SymbolEnv env) {
         if (expr != null && expr.getKind() == NodeKind.SIMPLE_VARIABLE_REF &&
                 (((BLangSimpleVarRef) expr).symbol.type.getKind() == TypeKind.FINITE ||
                 ((BLangSimpleVarRef) expr).symbol.type.getKind() == TypeKind.INTERSECTION)) {
             return ((BLangSimpleVarRef) expr).symbol.type; // Already type resolved constant
         }
+
+        type = Types.getReferredType(type);
+
         switch (type.tag) {
             case TypeTags.INT:
             case TypeTags.FLOAT:
             case TypeTags.DECIMAL:
                 BLangNumericLiteral numericLiteral = (BLangNumericLiteral) TreeBuilder.createNumericLiteralExpression();
-                return createFiniteType(constant, updateLiteral(numericLiteral, value, type, pos));
+                return createFiniteType(constantSymbol, updateLiteral(numericLiteral, value, type, pos));
             case TypeTags.BYTE:
                 BLangNumericLiteral byteLiteral = (BLangNumericLiteral) TreeBuilder.createNumericLiteralExpression();
-                return createFiniteType(constant, updateLiteral(byteLiteral, value, symTable.intType, pos));
+                return createFiniteType(constantSymbol, updateLiteral(byteLiteral, value, symTable.intType, pos));
             case TypeTags.STRING:
             case TypeTags.NIL:
             case TypeTags.BOOLEAN:
                 BLangLiteral literal = (BLangLiteral) TreeBuilder.createLiteralExpression();
-                return createFiniteType(constant, updateLiteral(literal, value, type, pos));
+                return createFiniteType(constantSymbol, updateLiteral(literal, value, type, pos));
             case TypeTags.MAP:
+            case TypeTags.RECORD:
                 if (value != null) {
-                    return createRecordType(expr, constant, value, pos, constValue);
+                    return createRecordType(expr, constantSymbol, value, pos, env);
+                }
+                return null;
+            case TypeTags.ARRAY:
+            case TypeTags.TUPLE:
+                if (value == null) {
```

Comment:
shall we make `value == null` check consistent with two paths. Between the group MAP, RECORD and group (ARRAY, TUPLE)